app/controllers/application_controller.rb

```diff
@@ -912,6 +913,13 @@ def add_noindex_header
     end
   end
 
+  def add_noindex_header_to_non_canonical
+    canonical = (@canonical_url || @default_canonical)
+    if canonical.present? && canonical != request.url && !SiteSetting.allow_indexing_non_canonical_urls
+      response.headers['X-Robots-Tag'] = 'noindex'
```

Comment:
```suggestion
      response.headers['X-Robots-Tag'] ||= 'noindex'
```
(or add `!response.headers['X-Robots-Tag']` to the `if` statement)